core-rust/node-common/src/scheduler.rs

```diff
@@ -62,190 +62,216 @@
-pub struct NoopScheduler;
-
-impl Scheduler for NoopScheduler {
-    type CancellationTokenType = NoopCancellableToken;
-
-    fn start_periodic(
-        &self,
-        _duration: Duration,
-        _task: impl 'static + FnMut() + Send,
-    ) -> Self::CancellationTokenType {
-        NoopCancellableToken {}
-    }
-}
-
-pub struct TokioScheduler {
-    runtime: Arc<Runtime>,
-}
-
-impl TokioScheduler {
-    pub fn new(runtime: Arc<Runtime>) -> Self {
-        Self { runtime }
+    ) -> RunningTask {
+        let cancellation_token = CancellationToken::new();
+        let running_task = RunningTask::new(cancellation_token.clone());
+        let Some(runtime) = self.runtime else {
+            // No runtime configured - we only fake the running, for test purposes:
+            return running_task;
+        };
+        // The `if` below seems larger than required - this is for "incompatible types" reasons
+        if let Some(metrics) = self.metrics {
+            let task = metrics.wrap(task, self.name.as_str());
+            Self::spawn_periodic_future(runtime, interval, task, cancellation_token)
+        } else {
+            Self::spawn_periodic_future(runtime, interval, task, cancellation_token)
+        };
+        if let Some(task_tracker) = self.running_task_tracker {
+            task_tracker.track(running_task.clone());
+        }
+        running_task
     }
-}
-
-impl Scheduler for TokioScheduler {
-    type CancellationTokenType = CancellationToken;
 
-    fn start_periodic(
-        &self,
-        duration: Duration,
+    fn spawn_periodic_future(
+        runtime: &Runtime,
+        interval: Duration,
         mut task: impl 'static + FnMut() + Send,
-    ) -> Self::CancellationTokenType {
-        let cancellation_token = Self::CancellationTokenType::new();
-        let cloned_token = cancellation_token.clone();
-        self.runtime.spawn(async move {
-            let mut interval = tokio::time::interval(duration);
-
+        cancellation_token: CancellationToken,
+    ) {
+        runtime.spawn(async move {
+            let mut interval = tokio::time::interval(interval);
             loop {
                 tokio::select! {
-                    _ = cloned_token.cancelled() => {
+                    _ = cancellation_token.cancelled() => {
                         break;
                     },
                     _ = interval.tick() => {
-                       task();
+                        task();
                     },
                 }
             }
         });
-        cancellation_token
     }
 }
 
-/// An auxiliary structure to be used along side a [`Scheduler`] in order to collect
-/// [`CancellableToken`]s for the spawned tasks in order to autostop them on drop.
-pub struct TaskTracker<CT: CancellableToken> {
-    tasks: Vec<CancelOnDrop<CT>>,
+/// A handle representing a running task started using the scheduler.
+/// Dropping *all* its clones cancels any subsequent executions of the task.
```

Comment:
_(note from author)_

I was considering whether we need an explicit `.cancel()` too (apart from "cancel on drop"). I didn't add it yet, to keep things simple (i.e. adding it in future is no big deal).